content/blog/similar-document-search-using-apache-spark-with-tf-idf.md

```diff
@@ -0,0 +1,167 @@
+
+```
+"authorDisplayName": ["Prasad Singathi","Maikel Pereira"],
+"publish": "2019-06-18T07:00:00.000Z",
+"category": ["machine-learning"],
+```
+
+- - -
+
+## Background
+
+As a professional services group, we were tasked with providing a solution to automatically find messages in the archives that are similar to new messages and send them to the person asking the question. 
+
+To accomplish that goal, we decided to apply machine learning to the process, so that there is an automated program able to find similarities between the current message and the historical data. The algorithm used was *term frequency&mdash;inverse document frequency* (TF-IDF). TF-IDF is used in a variety of applications. Typical use cases include document search, document tagging, and finding similar documents.  
+
+## Problem Description
+
+The desired solution was built using two Apache Spark applications running in a MapR cluster: one of them uses the historical data to update data features and train the model on a regular basis, and the second one analyzes every new message and finds five similar ones.
+
+## Application 1 - Creates Features and Trains Model
+
+This application was developed using Spark and Scala, and it can run on a schedule, depending on the needs. **Here is what it does, step by step:**
+
+1. Loads all messages from MapR Database. For the sake of brevity, we omit preprocessing steps like tokenization, stop words removal, punctuation removal, and other types of cleanup.
+
+```
+val rawEmaiData=spark.loadFromMapRDB("/googlegroups/messages")
+val rawEmaiDataDF=rawEmaiData.select("_id","bodyWithHistory","threadId","emailDate")
+```
+
+2. Creates hashingTF, using HashingTF class available in Spark, and sets fixed-length feature vectors of 1000. It applies the hashing transformation to the document, resulting in the featurizedData.
+
+```
+val hashingTF = new HashingTF().setInputCol("words").setOutputCol("rawFeatures").setNumFeatures(1000)
+val featurizedData = hashingTF.transform(wordsData)
+```
+
+3. Creates the IDF, and from the TF and the IDF, it creates the TF-IDF.
+
+```
+val idf = new IDF().setInputCol("rawFeatures").setOutputCol("features")
+val idfModel = idf.fit(featurizedData)
+val rescaledData = idfModel.transform(featurizedData)
+```
+
+4. A UDF is necessary for pre-calculating sparse vector norm.
+
+```
+def calcNorm(vectorA: SparseVector): Double = {
+      var norm = 0.0
+      for (i <-  vectorA.indices){ norm += vectorA(i)*vectorA(i) }
+      (math.sqrt(norm))
+    }
+val calcNormDF = udf[Double,SparseVector](calcNorm)
+```
+
+5. Creates a TF-IDF corpus.
+
+```
+val normalized = rescaledData.withColumn("norm",calcNormDF(col("features")))
+```
+
+6. Saves IDF model to MapR XD Distributed File and Object Store.
+
+```
+idfModel.write.overwrite().save("/googlegroups/save_model_idf")
+```
+
+7. To save features vector to the MapR Database table, we have to convert the features vector to JSON format; for this, we create and register a UDF.
```

Comment:
Recommend changing the semi colon to a period and starting the next sentence with "For this, we create..."